### PR TITLE
Skip codegen for eponymous templates aliasing a type

### PR DESCRIPTION
**Pull request by kinke**
Such as `{core.internal,std}.traits.Unqual!T`.

Also perform the discard check before recursively traversing the potentially huge graph of parents and siblings in `needsCodegen()`.

### CODE REVIEW

**Comment by kinke** [comment]
[This fixes a `needsCodegen()` stack overflow for a Symmetry project.]

**Comment by andralex** [comment]
Added a new label `industry`

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -6125,55 +6125,80 @@ extern (C++) class TemplateInstance : ScopeDsymbol
...
-            return false;
+        if (auto v = aliasdecl.isVarDeclaration())
+        {
+            /** Detect & discard:
+                ---
+                template foo(args...) {
+                    some_basic_type_or_string helper() { .... }
+                    enum foo = helper();
+                }
+                ---
```
Comment (WalterBright) [inline src/dmd/dtemplate.d:6142]: These --- lines are for Ddoc and are quite unnecessary here.

File: src/dmd/dtemplate.d [hunk 2]
```diff
@@ -6125,55 +6125,80 @@ extern (C++) class TemplateInstance : ScopeDsymbol
...
+                template foo(args...) {
+                    import bar : Symbol;
+                    alias foo = Symbol!args;
+                }
+                ---
+            */
+
+            // Only *type* aliases
+            if (a.aliassym)
```
Comment (WalterBright) [inline src/dmd/dtemplate.d:6179]: This doesn't just trigger on type aliases. a.aliassym can be symbol, like a function symbol.

File: src/dmd/dtemplate.d [hunk 3]
```diff
@@ -6125,55 +6125,80 @@ extern (C++) class TemplateInstance : ScopeDsymbol
...
+
+            // Only allow the eponymous member + import members, as the aliased type might be
+            // an inner aggregate requiring codegen.
+            size_t numNonImportMembers;
+            foreach (member; *members)
+                if (!member.isImport())
+                    numNonImportMembers++;
+
+            assert(numNonImportMembers >= 1);
+            if (numNonImportMembers > 1)
```
Comment (WalterBright) [inline src/dmd/dtemplate.d:6190]: this test can be moved into the loop.

File: src/dmd/dtemplate.d [hunk 4]
```diff
@@ -6125,55 +6125,80 @@ extern (C++) class TemplateInstance : ScopeDsymbol
...
+            */
+
+            // Only *type* aliases
+            if (a.aliassym)
                 return false;
-            if(member.isStaticCtorDeclaration())
+
+            // Only allow the eponymous member + import members, as the aliased type might be
+            // an inner aggregate requiring codegen.
+            size_t numNonImportMembers;
```
Comment (WalterBright) [inline src/dmd/dtemplate.d:6184]: This only needs to count to 2. Make it `uint`.

**Comment by WalterBright** [comment]
@andralex you wrote this fixes a compiler stack overflow. Please add a bugzilla for the stack overflow, and make this the fix for the bugzilla. Also, this PR should explain what it is doing that avoids the stack overflow.

**Comment by kinke** [comment]
This isn't a real fix, just a tentative `needsCodegen()` optimization for trivially type-aliasing templates. The problem was encountered for `Unqual` (it probably doesn't help that each Phobos `Unqual!T` instantiates a druntime `Unqual!T`, presumably for the sake of not globally importing `core.internal.traits` in `std.traits`).

As the linker errors show, the constraints for this optimization currently aren't restrictive enough; it depends on the aliased type not instantiating something new which isn't instantiated on the outside, like the failing `core.atomic.TailShared` template: https://github.com/dlang/druntime/blob/4ca9c36b91af06d92d69e94e72d3c77395409d20/src/core/atomic.d#L713-L774

Probably related: https://issues.dlang.org/show_bug.cgi?id=18026 and https://issues.dlang.org/show_bug.cgi?id=21473

File: src/dmd/dtemplate.d (hunk 3 above)
Comment (UplinkCoder) [reply src/dmd/dtemplate.d:6190]: why would you move it into the loop?

**Comment by RazvanN7** [comment]
@kinke are you planning on pursuing this further?

**Comment by kinke** [comment]
Nope, I don't think it looks promising anymore when having to add a check for the case above - closing.

**Pull request closed without merging**